### PR TITLE
Refactor Network Layer

### PR DESCRIPTION
**Pull request by germanStabile**
#### Description:

* Rollback to previous networking implementation, removing Moya. And upgrading Alamofire to ~> 5.0

### CODE REVIEW

File: ios-base/Onboarding/ViewModels/FirstViewModel.swift [hunk 1]
```diff
@@ -43,15 +43,16 @@ class FirstViewModel {
...
-    UserService.sharedInstance.loginWithFacebook(
-      token: token.tokenString,
-      success: { [weak self] in
-        self?.state = .network(state: .idle)
-        AppNavigator.shared.navigate(to: HomeRoutes.home, with: .changeRoot)
-      },
-      failure: { [weak self] error in
-        self?.state = .network(state: .error(error.localizedDescription))
-    })
+    //TODO: remove when implementing login with apple id
```
Comment (pMalvasio) [inline ios-base/Onboarding/ViewModels/FirstViewModel.swift:46]: I would say, remove it now. No need to let Facebook as dependency there if in the near future Apple Sign-in will be there

Comment (germanStabile) [reply ios-base/Onboarding/ViewModels/FirstViewModel.swift:46]: I can do that in a follow-up PR, this one is pretty big already

Comment (pMalvasio) [reply ios-base/Onboarding/ViewModels/FirstViewModel.swift:46]: Cool. No need to do it here. But for sure no need to wait until Apple Sign-in is ready

File: ios-base/Networking/Services/APIClient.swift [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+  //If your API requires this header
+  //do not use this method or change backend to skip this validation.
+  class func multipartRequest(_ method: HTTPMethod = .post,
+                              url: String,
+                              headers: [String: String] = APIClient.getHeaders(),
+                              params: [String: Any]?,
+                              paramsRootKey: String,
+                              media: [MultipartMedia],
+                              success: @escaping SuccessCallback,
+                              failure: @escaping FailureCallback) {
```
Comment (glm4) [inline ios-base/Networking/Services/APIClient.swift:153]: let's use the new indentation style 😃

File: ios-base/Networking/Services/APIClient.swift [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+    
+    return request
+  }
+  
+  init(path: String,
+       baseUrl: String = APIClient.getBaseUrl(),
+       method: HTTPMethod,
+       encoding: ParameterEncoding? = nil,
+       params: [String: Any]? = nil,
+       headers: [String: String] = [:]) {
```
Comment (glm4) [inline ios-base/Networking/Services/APIClient.swift:61]: Use the new indentation style

File: ios-base/Networking/Services/APIClient.swift [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+
+class BaseURLRequestConvertible: URLRequestConvertible {
+  let url: URLConvertible
+  let method: HTTPMethod
+  let headers: HTTPHeaders
+  let params: [String: Any]?
+  let encoding: ParameterEncoding?
+  
+  func asURLRequest() throws -> URLRequest {
+    let request = try URLRequest(url: url,
```
Comment (glm4) [inline ios-base/Networking/Services/APIClient.swift:46]: Use the new indentation style from the guide

File: ios-base/Networking/Services/APIClient.swift [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+    
+    AF.upload(
+      multipartFormData: { (multipartForm) -> Void in
+        if let parameters = params {
+          multipartFormData(multipartForm, params: parameters, rootKey: paramsRootKey)
+        }
+        for elem in media {
+          elem.embed(inForm: multipartForm)
+        }
+    },
```
Comment (glm4) [inline ios-base/Networking/Services/APIClient.swift:169]: this } should be aligned with `multipartFormData` right?

File: ios-base/Networking/Services/APIClient.swift [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+      with: requestConvertible)
```
Comment (glm4) [inline ios-base/Networking/Services/APIClient.swift:170]: I will close the method call in a new line so the subsequent call don't appear at the same level that the previous call parameter

File: ios-base/Networking/Services/APIClient.swift [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+  class func defaultEncoding(forMethod method: HTTPMethod) -> ParameterEncoding {
+    switch method {
+    case .post, .put, .patch:
+      return JSONEncoding.default
+    default:
+      return URLEncoding.default
+    }
+  }
+  
+  class func request(_ method: HTTPMethod,
```
Comment (glm4) [inline ios-base/Networking/Services/APIClient.swift:185]: same for indentation style

File: ios-base/Networking/Services/APIClient.swift [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+    failure: @escaping FailureCallback
+  ) -> Bool {
+    let defaultError = App.error(
+      domain: .generic,
+      localizedDescription: "Error parsing response".localized
+    )
+    
+    guard
+      let data = data,
+      !data.isEmpty
```
Comment (glm4) [inline ios-base/Networking/Services/APIClient.swift:292]: Please don't split lines if not needed.

File: ios-base/Networking/Services/APIClient.swift [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+  
+  fileprivate class func validateEmptyResponse(
+    response: HTTPURLResponse,
+    data: Data?,
+    success: @escaping SuccessCallback,
+    failure: @escaping FailureCallback
+  ) -> Bool {
+    let defaultError = App.error(
+      domain: .generic,
+      localizedDescription: "Error parsing response".localized
```
Comment (glm4) [inline ios-base/Networking/Services/APIClient.swift:287]: maybe a more suitable error for this case would be "Unexpected empty response" or something like that.

File: ios-base/Networking/Services/APIClient.swift [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+    return nil
+  }
+  
+  fileprivate class func validateResult(
+    result: AFDataResponse<Any>,
+    success: @escaping SuccessCallback,
+    failure: @escaping FailureCallback
+  ) {
+    let defaultError = App.error(
+      domain: .generic,
```
Comment (glm4) [inline ios-base/Networking/Services/APIClient.swift:249]: Shouldn't we use the parsing domain?

File: ios-base/Networking/Services/APIClient.swift [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+    failure: @escaping FailureCallback
+  ) {
+    
+    var dictionary: [String: Any]?
+    var serializationError: NSError?
+    do {
+      dictionary = try JSONSerialization.jsonObject(
+        with: data,
+        options: .allowFragments
+        ) as? [String: Any]
```
Comment (glm4) [inline ios-base/Networking/Services/APIClient.swift:319]: align with `dictionary`

File: ios-base/Networking/Services/AuthenticationServices.swift [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+import Foundation
+import UIKit
+
+class AuthenticationServices {
+  
+  fileprivate static let usersUrl = "/users/"
+  fileprivate static let currentUserUrl = "/user/"
+  
+  class func login(_ email: String,
```
Comment (glm4) [inline ios-base/Networking/Services/AuthenticationServices.swift:17]: new indentation style

File: ios-base/Networking/Services/AuthenticationServices.swift [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+      AuthenticationServices.saveUserSession(fromResponse: response, headers: headers)
+      success()
+    }, failure: { error in
+      failure(error)
+    })
+  }
+  
+  //Multi part upload example
+  //TODO: rails base backend not supporting multipart uploads yet
+  class func signup(_ email: String,
```
Comment (glm4) [inline ios-base/Networking/Services/AuthenticationServices.swift:38]: same here for indentation

File: ios-base/Networking/Services/AuthenticationServices.swift [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+    //Then: media [image, image2]
+    APIClient.multipartRequest(
+      url: usersUrl,
+      params: parameters,
+      paramsRootKey: "",
+      media: [image],
+      success: { response, headers in
+        AuthenticationServices.saveUserSession(fromResponse: response, headers: headers)
+        success(UserDataManager.currentUser)
+    },
```
Comment (glm4) [inline ios-base/Networking/Services/AuthenticationServices.swift:70]: align with `success` alias

File: ios-base/Networking/Services/AuthenticationServices.swift [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+    ]
+    
+    APIClient.request(
+      .post,
+      url: usersUrl,
+      params: parameters,
+      success: { response, headers in
+        AuthenticationServices.saveUserSession(fromResponse: response, headers: headers)
+        success(UserDataManager.currentUser)
+    },
```
Comment (glm4) [inline ios-base/Networking/Services/AuthenticationServices.swift:98]: same here

File: ios-base/Common/Models/User.swift [hunk 16]
```diff
@@ -20,4 +20,11 @@ struct User: Codable {
     case email
     case image = "profile_picture"
   }
+    
+  init(dictionary: [String: Any]) {
+    self.id = dictionary[CodingKeys.id.rawValue] as? Int ?? 0
+    self.username = dictionary[CodingKeys.username.rawValue] as? String ?? ""
+    self.email = dictionary[CodingKeys.email.rawValue] as? String ?? ""
+    self.image = URL(string: dictionary[CodingKeys.image.rawValue] as? String ?? "")
+  }
```
Comment (glm4) [inline ios-base/Common/Models/User.swift:29]: maybe we can make this an optional initializer, so we don't allow creating 'empty' user objects

**Comment by germanStabile** [comment]
> PD: I'd move the implementation using Moya to a separate branch with a disclaimer that we are not going to maintain it.

https://github.com/rootstrap/ios-base/tree/feature/moya_integration

File: ios-base/Networking/Services/UserServices.swift [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+//  UserServices.swift
+//  ios-base
+//
+//  Created by Germán Stábile on 6/8/20.
+//  Copyright © 2020 TopTier labs. All rights reserved.
+//
+
+import Foundation
+
+class UserServices {
```
[the remark this answers is not in the document]
Comment (germanStabile) [reply ios-base/Networking/Services/UserServices.swift:11]: I don't know about this one, I kinda like it as is.

**Comment by germanStabile** [comment]
CI succeeded not sure why it hanged expecting status report

**Pull request merged**